Reorganize Security Best Practices into Better Categories (#3034)
dfinity/portal: docs/developer-docs/security/security-best-practices/overview.mdx

---

> sidebar_position: 1
> sidebar_label: Overview

Reviewer: ```suggestion

```

Author: Done

keywords: [security, concept]
---

import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";

> # Security best practices: Overview

Reviewer: ```suggestion
# Overview
```

Author: Done

<MarkdownChipRow labels={["Intermediate", "Security", "Concept"]} />

> This document provides security best practices for developing canisters and web apps served by canisters on the Internet Computer. These best practices are mostly inspired by issues found in security reviews.

Author: Done

The goal of these best practices is to enable developers to identify and address potential issues early during development of new dapps, and not only in the end when (if at all) a security review is done. Ideally, this will make the development of secure dapps more efficient.

> Some excellent canister best practices linked here are from [effective Rust canisters](https://mmapped.blog/posts/01-effective-rust-canisters.html) and [how to audit an Internet Computer canister](https://www.joachim-breitner.de/blog/788-How_to_audit_an_Internet_Computer_canister). The relevant sections are linked in the individual best practices.

Reviewer: ```suggestion
Some excellent canister best practices linked here are from [effective Rust canisters](https://mmapped.blog/posts/01-effective-rust-canisters.html) and [how to audit an ICP canister](https://www.joachim-breitner.de/blog/788-How_to_audit_an_Internet_Computer_canister). The relevant sections are linked in the individual best practices.
```

Author: Done

## Target audience

> This document was initially intended for internal use at DFINITY. However, it has now been published to the community so every developer can benefit. The target audience are any developers working on canisters or web apps for the Internet Computer. This is also of interest to anyone doing reviews of such code.

Author: Thanks, done with slight modifications

## Disclaimers and limitations

> The collection of best practices may grow over time. While it is useful to improve security of dapps on the Internet Computer, such a list will never be complete and will never cover all potential security concerns. For example, there will always be attack vectors very specific to a dapps use cases that cannot be covered by general best practices. Thus, following the best practices can complement, but not replace security reviews. Especially for security critical dapps it is recommended to perform security reviews/audits. Furthermore, please note that the best practices are currently not ordered according to risk or priority.

Reviewer 2: I wouldn't focus on security reviews and audits as much.

These security best practices complement other security activities in the SDLC. Critical dapps need a higher level of maturity on all secure software development activities.

We actually say that earlier 

> The goal of these best practices is to enable developers to identify and address potential issues early during development of new dapps, and not only in the end when (if at all) a security review is done.

But then here we only focus on the best practices and the reviews. I always like [this visual](http://www.swsec.com/resources/touchpoints/) to see what more you should be doing.

Author: Fully agree, as discussed, I'll create a subsequent PR to adapt this as this PR is only moving this text.

> The collection of best practices may grow over time. While it is useful to improve security of dapps on the Internet Computer, such a list will never be complete and will never cover all potential security concerns. For example, there will always be attack vectors very specific to a dapps use cases that cannot be covered by general best practices. Thus, following the best practices can complement, but not replace security reviews. Especially for security critical dapps it is recommended to perform security reviews/audits. Furthermore, please note that the best practices are currently not ordered according to risk or priority.

Reviewer: ```suggestion
The collection of best practices may grow over time. While it is useful to improve security of dapps on ICP, such a list will never be complete and will never cover all potential security concerns. For example, there will always be attack vectors very specific to a dapps use cases that cannot be covered by general best practices. Thus, following the best practices can complement, but not replace security reviews. Especially for security critical dapps, it is recommended to perform security reviews or audits. Furthermore, please note that the best practices are currently not ordered according to risk or priority.
```

Author: Done
